### PR TITLE
rustc_metadata: Privatize more things and a couple of other refactorings

### LINKED ISSUE #66550: AllocatorKind::GlobalExe is never constructed
Defined at https://github.com/rust-lang/rust/blob/9285d401a6070094747465962bc49969b93e14c5/src/libsyntax/expand/allocator.rs#L9

Search results: https://github.com/rust-lang/rust/search?q=DefaultExe

### PR DESCRIPTION
**Pull request by petrochenkov**
This PR continues https://github.com/rust-lang/rust/pull/66496 and hits the point of diminishing returns.
All fields of `CrateRoot` and `CrateMetadata` are privatized.
For read-only fields this certainly makes sense, but for a few fields updateable from outside of `rmeta.rs` (mostly `creader.rs`) it was done mostly for consistency, I can make them `pub(crate)` again if requested.

`cstore.rs` (which became small after #66496) was merged into `creader.rs`.

A few things noticed while making the privacy changes were addressed in the remaining refactoring commits.

Fixes https://github.com/rust-lang/rust/issues/66550
r? @eddyb @Mark-Simulacrum

### CODE REVIEW

File: src/libsyntax/expand/allocator.rs [hunk 1]
```diff
@@ -6,15 +6,13 @@ use syntax_pos::Span;
 pub enum AllocatorKind {
     Global,
     DefaultLib,
```
Comment (bjorn3) [inline src/libsyntax/expand/allocator.rs]:
```suggestion
    Default,
```

**Review by eddyb — APPROVED** [review]

File: src/librustc/ty/context.rs [hunk 2]
```diff
@@ -1307,10 +1307,6 @@ impl<'tcx> TyCtxt<'tcx> {
         self.all_crate_nums(LOCAL_CRATE)
     }
 
-    pub fn injected_panic_runtime(self) -> Option<CrateNum> {
-        self.cstore.injected_panic_runtime()
-    }
```
Comment (RalfJung) [inline src/librustc/ty/context.rs]: Miri was relying on this method... is there a replacement?

Comment (petrochenkov) [reply src/librustc/ty/context.rs]: Hmm, something like
```rust
tcx.crates().iter().find(|cnum| tcx.is_panic_runtime(cnum))
```
should work without re-exposing things privatized in this PR.

(If that's not ok, I can send a PR re-adding `tcx.injected_panic_runtime()`.)

Comment (RalfJung) [reply src/librustc/ty/context.rs]: That seems to work, thanks! (modulo some extra `*`)

See https://github.com/rust-lang/miri/pull/1085

**Pull request merged**